### ISSUE TITLE
Fix S4158 FP: collection filled by another function

### ISSUE THREAD

**Issue by CristianAmbrosini** [SonarSource/sonar-dotnet#9444]
Source: https://community.sonarsource.com/t/false-positive-for-s4158/117480

S4158 is incorrectly raised when a list is filled by another function and then iterated:
```cs
class CollectionFilledByAnotherFunction
{
  List<int> Numbers { get; set; }

  void Initialize()
  {
    Numbers = new List<int>();
    InitializeNumbers();
    foreach (var number in Numbers) // Noncompliant FP
    {
      System.Console.WriteLine(number);
    }
  }

  void InitializeNumbers()
  {
    Numbers.Add(1);
    Numbers.Add(2);
    Numbers.Add(3);
  }
}
```

**Comment by Tim-Pohlmann** [comment]
The cause of the problem is that the invocation processor is only resetting field states. Properties are missing from the logic:
https://github.com/SonarSource/sonar-dotnet/blob/f563caea9f54418f67757230d30fb3636000e6c7/analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.cs#L48-L51
